### DISCUSSION TITLE
msw/node 'headers-polyfill' error with Vite >= 4.1.0

### DISCUSSION THREAD

**Discussion by jgarplind** [mswjs/msw#1547] [Q&A] [answered]
Been running Vite with msw/node for our unit tests for a while and everything has been working well, but upgrading to Vite 4.10 (which internally upgraded rollup to >= 3.10) has caused this issue to block test execution:

> SyntaxError: Named export 'Headers' not found. The requested module 'headers-polyfill' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:
import pkg from 'headers-polyfill';
const { Headers } = pkg;

I tried reproducing the error with the issue template but failed initially, so before committing more time to reproducing the issue, I wanted to check in with the community if this error is discernable and maybe even patchable without deeper investigation?

Thanks for any inputs.

**Comment by kettanaito** [comment, marked as answer]
Hey, @jgarplind. It looks like you're using MSW in an ESM project while MSW doesn't support ESM at the moment. Its dependency, `headers-polyfill`, is also CJS-only, so it causes the error above. 

I think this should be addressed as a part of our ESM support, which is in progress.

**Reply by jgarplind** [reply]
Thanks for clarifying!

I can't say I'm too well-versed with the differences between ESM and CJS, and it seems to be a never-ending debacle, so I'll just accept the status quo and keep Vite slightly outdated for now.

Still a bit confused about why everything has been running fine up until Vite 4.0.4 but not using Vite 4.1.0 [Changelog](https://github.com/vitejs/vite/blob/main/packages/vite/CHANGELOG.md#410-2023-02-02). It doesn't happen to be something you understand from a cursory glance?

Thanks again.

**Reply by mattcosta7** [reply]
> Vite 4.1 updates to the latest versions of Rollup and esbuild. 

I think it might be related to how roll-up and esbuild transpile in 4.1? 

I haven't looked into vite in a while, but that would be my best guess - something is transpiling differently now 

Maybe there's an interop option in roll-up not being used, that was previously?

**Reply by jgarplind** [reply]
Maybe! Thanks for offering helpful ideas.

Another interesting input, possibly:

While trying to find the exact commit that caused the changed behavior, I tried installing my dependencies with `pnpm` instead of `npm` in order to be able to run Vite from a local folder (Vite blocks usage of `npm` using https://pnpm.io/only-allow-pnpm it seems).

Unexpectedly, using `pnpm i` instead of `npm i` actually ending up making things work (like before).

I think this may just be another reason for us to move to `pnpm`, but I can't say I'm any closer to understanding the underlying issue.

**Reply by jdk2pq** [reply]
I was having the same issue and was already using `pnpm`. 

But I figured out that if I deleted my `pnpm-lock.yml` file and did the ol' `rm -rf node_modules` and reinstalled everything from scratch with `pnpm i`, things started working again. 🤷‍♂️

Hope it helps someone else!
